File: src/pcapack/src/utils/distances.f90
! This Source Code Form is subject to the terms of the Mozilla Public
! License, v. 2.0. If a copy of the MPL was not distributed with this
! file, You can obtain one at http://mozilla.org/MPL/2.0/.

! Copyright 2013-2014, Schmidt


module distances
  use :: lapack
  implicit none
  
  integer, public, parameter :: dist_euclidean = 1
  integer, public, parameter :: dist_supremum  = 2
  integer, public, parameter :: dist_infimum   = 3
  integer, public, parameter :: dist_manhattan = 4
  integer, public, parameter :: dist_rcanberra = 5
  integer, public, parameter :: dist_minkowski = 6
  integer, public, parameter :: dist_canberra  = 7
  !  integer, public, parameter :: dist_binary    = 8
  
  
  contains
  
  
  function distance(method, n, x, y, p) result(dist)
    double precision :: dist
    ! in/out
    integer, intent(in) :: method, n, p
    double precision, intent(in) :: x(n), y(n)
    ! local
    integer :: i, allocerr
    integer :: ux, uy
    double precision, allocatable :: work(:)
    
    
    allocerr = 0
    allocate(work(n), stat=allocerr)
    if (allocerr /= 0) stop "out of memory"
    
    ! work = x - y
    call dcopy(n, x, 1, work, 1)
    
    call daxpy(n, -1.0d0, y, 1, work, 1)
    
    if (method == dist_euclidean) then
      dist = dnrm2(n, work, 1)
    else if (method == dist_supremum) then
      dist = work(idamax(n, work, 1))
    else if (method == dist_infimum) then
      dist = work(idamin(n, work, 1))
      return
    else if (method == dist_manhattan) then
      dist = dasum(n, work, 1)
    else if (method == dist_rcanberra) then
      dist = 0.0d0
      do i = 1, n
        dist = dist + abs(work(i)) / abs(x(i-1) + y(i-1))
      end do
    else if (method == dist_canberra) then
      dist = 0.0d0
      do i = 1, n
        dist = dist + abs(work(i)) / (abs(x(i-1)) + abs(y(i-1)))
      end do
    else if (method == dist_minkowski) then
      dist = dnrm3(n, work, 1, p)
!    else if (method == dist_binary) then
!      dist = 0.0d0
!      do i = 1, n
!        ux = transfer(x(i), 1)
!        uy = transfer(y(i), 1)
!        
!        
!      end do
    end if
    
    deallocate(work)
    
    return
  end function
  
  
  
  ! mahalanobis distance d_m(x,y)
  function distance_mahalanobis(n, x, y, covinv) result(mahal)
    double precision :: mahal
    ! in/out
    integer, intent(in) :: n
    double precision, intent(in) :: x(n), y(n), covinv(*)
    ! local
    integer :: i, allocerr
    double precision, allocatable :: work1(:)
    double precision, allocatable :: work2(:)
    
    
    allocerr = 0
    allocate(work1(n), stat=allocerr)
    if (allocerr.ne.0) stop "out of memory"
    allocate(work2(n), stat=allocerr)
    if (allocerr.ne.0) stop "out of memory"
    
    ! work1 = x - y
    call dcopy(n, x, 1, work1, 1)
    
!    call daxpy(n, 1.0d0, work1, 1, y, 1)
    
    ! mahal = work1^t * covinv * work2
    call dgemv('n', n, n, 1.0d0, covinv, n, work1, 1, 0.0d0, work2, 1)
    
    mahal = sqrt(ddot(n, work1, 1, work2, 1))
    
    deallocate(work1)
    deallocate(work2)
    
    return
  end function
  
  
  
!  subroutine distance_matrix(method, m, n, x, ret, p)
!    integer, intent(in) :: method, n, p
!    double precision, intent(in) :: x(m, n), ret(*)
!    integer :: retlen
!    
!    
!    retlen = n*(n-1)/2
!    
!    
!    
!  end subroutine
end module

File: src/pcapack/src/matlib/distances.c
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

// Copyright 2015, Schmidt

#include <stdlib.h>
#include <math.h>

#include "matlib.h"
#include "lapack.h"
#include "../omp.h"


static inline int idamax(const int n, const double *restrict x)
{
  int i;
  int ret = 0;
  double max = fabs(x[0]);
  
  #pragma omp parallel for simd
  for (i=1; i<n; i++)
  {
    if (fabs(x[i]) > max)
    {
      max = x[i];
      ret = i;
    }
  }
  return ret;
}

static inline int idamin(const int n, const double *restrict x)
{
  int i;
  int ret = 0;
  double min = fabs(x[0]);
  
  #pragma omp parallel for simd default(shared)
  for (i=1; i<n; i++)
  {
    if (fabs(x[i]) < min)
    {
      min = x[i];
      ret = i;
    }
  }
  return ret;
}

// Modification of DNRM2 for more general Minkowski 
// Original code is:
// This version written on 25-October-1982.
//     Modified on 14-October-1993 to inline the call to DLASSQ.
//     Sven Hammarling, Nag Ltd.
// Modified code by Drew Schmidt, 2015
static double dnrm3(const int n, double *restrict x, const int p)
{
  int i;
  const double dp = (double) p;
  double absxi;
  double scale = 0., ssq = 1.;
  double nrm = 0.;
  
  if (n < 1) 
    return 0.;
  else if (n == 1) 
    return fabs(x[0]);
  

  for (i=0; i<n; i++)
  {
    if (x[i] != 0)
    {
      absxi = fabs(x[i]);
      if (scale < absxi)
      {
        ssq = 1. + ssq*pow(scale/absxi, dp);
        scale = absxi;
      }
      else
        ssq += pow(absxi/scale, dp);
    }
  }
  
  nrm = scale * pow(ssq, 1./dp);
  return nrm;
}



double pcapack_distance(const int method, const int n, const double *restrict x, const double *restrict y, const int p)
{
  double dist = 0;
  double tmp;
  int i;
  int ux, uy;
  double *work;
  
  int i1 = 1;
  double neg1 = -1.;
  
  work = malloc(n * sizeof(*work));
  
  
  SAFE_PARALLEL_FOR_SIMD
  for (i=0; i<n; i++)
    work[i] = x[i] - y[i];
  
  
  if (method == DIST_EUCLIDEAN)
    dist = dnrm2(n, work, 1);
  else if (method == DIST_SUPREMUM)
    dist = work[idamax(n, work)];
  else if (method == DIST_INFIMUM)
    dist = work[idamin(n, work)];
  else if (method == DIST_MANHATTAN)
  {
    #pragma omp parallel for simd default(shared) reduction(+:dist)
    for (i=0; i<n; i++)
      dist += fabs(work[i]);
  }
  else if (method == DIST_CANBERRA)
  {
    #pragma omp parallel for simd default(shared) reduction(+:dist)
    for (i=0; i<n; i++)
    {
      tmp = fabs(x[i]) + fabs(y[i]);
      if (tmp != 0.)
        dist += fabs(work[i]) / tmp;
    }
  }
  else if (method == DIST_RCANBERRA)
  {
    #pragma omp parallel for simd default(shared) reduction(+:dist)
    for (i=0; i<n; i++)
    {
      tmp = fabs(x[i] + y[i]);
      if (tmp != 0.)
        dist += fabs(work[i]) / tmp;
    }
  }
  else if (method == DIST_MINKOWSKI)
    dist = dnrm3(n, work, p);
  // The distance is the _proportion_ of bits in which only one is
  // on amongst those in which at least one is on.
  else if (method == DIST_BINARY)
  {
    int on = 0;
    
    #pragma omp parallel for simd default(shared) reduction(+:dist,on)
    for (i=0; i<n; i++)
    {
      if (x[i] || y[i])
      {
        on++;
        if (!(x[i] && y[i]))
          dist += 1.;
      }
    }
    
    if (on != 0)
      dist /= (double) on;
  }
  
  free(work);
  return dist;
}



double pcapack_mahalanobis(int n, const double *restrict x, const double *restrict y, const double *restrict covinv)
{
  int i;
  double mahal = 0.;
  double *work1, *work2;
  
  work1 = malloc(n * sizeof(*work1));
  work2 = malloc(n * sizeof(*work2));
  
  SAFE_PARALLEL_FOR_SIMD
  for (i=0; i<n; i++)
    work1[i] = x[i] - y[i];
  
  // mahal = work1^t * covinv * work2
  dgemv_(&(char){'n'}, &n, &n, &(double){1.}, covinv, &n, work1, &(int){1}, &(double){0.}, work2, &(int){1});
  
  #pragma omp parallel for simd default(shared) reduction(+:mahal)
  for (i=0; i<n; i++)
    mahal += work1[i]*work2[i];
  
  mahal = sqrt(mahal);
  
  free(work1);
  free(work2);
  return mahal;
}

